operator/controllers/instance_controller.go

```diff
@@ -0,0 +1,461 @@
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// TODO: update the status with validated so the controller doesn't keep performing the
+	// validation on requeues.
+	if *instance.Spec.ValidateValues {
+		schemaJSON, err := yaml.YAMLToJSON([]byte(plan.Spec.Provisioning.Values.Schema))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		schema := gojsonschema.NewBytesLoader(schemaJSON)
+		valuesJSON, err := yaml.YAMLToJSON([]byte(instance.Spec.Values))
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		values := gojsonschema.NewBytesLoader(valuesJSON)
+		result, err := gojsonschema.Validate(schema, values)
+		if err != nil {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values.
+			return ctrl.Result{}, nil
+		}
+		if !result.Valid() {
+			log.Error(err, "failed to validate values")
+			// TODO: update status with invalid values. Including specific errors returned from
+			// result.Errors().
+			return ctrl.Result{}, nil
+		}
+	}
+
+	valuesSecretName := fmt.Sprintf("metabroker-%s-values", instance.Name)
+	if created, err := r.valuesSecret(ctx, instance, valuesSecretName, instance.Namespace); err != nil {
+		return ctrl.Result{}, err
+	} else if created {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	helmInstanceName := fmt.Sprintf("metabroker-%s", instance.Name)
+	podName := fmt.Sprintf("metabroker-%s-provision", instance.Name)
+	return r.provisioningPod(ctx, instance, plan, helmInstanceName, podName, instance.Namespace, valuesSecretName)
+}
+
+func (r *InstanceReconciler) setOwnership(
+	ctx context.Context,
+	instance *servicebrokerv1alpha1.Instance,
+	plan *servicebrokerv1alpha1.Plan,
+) error {
+	if err := ctrl.SetControllerReference(plan, instance, r.Scheme); err != nil {
+		return fmt.Errorf("failed to set ownership: %w", err)
+	}
+	return nil
+}
+
+// valuesSecret creates a Secret containing the values.yaml content if it doesn't exist yet.
+// The first return parameter represents whether the Secret was created or not.
+func (r *InstanceReconciler) valuesSecret(
+	ctx context.Context,
+	instance *servicebrokerv1alpha1.Instance,
+	valuesSecretName string,
+	namespace string,
+) (bool, error) {
+	namespacedName := types.NamespacedName{
+		Name:      valuesSecretName,
+		Namespace: namespace,
+	}
+	current := &corev1.Secret{}
+	if err := r.Get(ctx, namespacedName, current); err != nil {
+		if !errors.IsNotFound(err) {
+			return false, fmt.Errorf("failed to process values secret: %w", err)
+		}
+		desired := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      namespacedName.Name,
+				Namespace: namespacedName.Namespace,
+				// TODO: add proper labels.
+			},
+			Data: map[string][]byte{"values.yaml": []byte(instance.Spec.Values)},
```

Comment:
From interacting with quarks, I've found that keys with dots in them kinda suck (because jsonpath, jq, etc. tend to use dots for field separators).  But (I haven't checked yet) if we're mounting this as a file somewhere I guess that's just the way it is.

Reply:
That's the case, we mount as a file. I could do a re-mapping, but escaping the dot should also be straightforward? I don't recall having issues with dots myself.